File: javascript/leet_code/881.mjs
/**
 * @file This module is responsible for calculating the minimum number of rescue boats required to
 * save people.
 * Each boat carries a weight limit, and each person has a weight. The goal is to minimize the
 * number of boats needed.
 * The solution uses a two-pointer technique to pair the heaviest and lightest person until all
 * people are accounted for.
 * @module leet_code/881
 */

/**
 * Calculates the minimum number of rescue boats required.
 *
 * The function sorts the array of people's weights in ascending order. Then, it uses a two-pointer
 * technique to pair the heaviest person (right pointer) with the lightest person (left pointer) if
 * possible. If the sum of their weights is less than or equal to the limit, they can share a boat,
 * and the left pointer moves to the next person. If not, the heaviest person takes a boat alone,
 * and the right pointer moves to the next person. This process continues until all people are
 * accounted for.
 *
 * @param {number[]} people - An array of integers where each integer represents the weight of a person.
 * @param {number} limit - The maximum weight that a rescue boat can carry.
 * @returns {number} The minimum number of rescue boats required.
 */
const numRescueBoats = function (people, limit) {
    // Initialize the result to 0
    let result = 0;
    // Initialize two pointers, left and right
    let l = 0, r = people.length - 1;
    // Sort the array of people's weights in ascending order
    people.sort((a, b) => a - b);
    // Continue until all people are accounted for
    while (l <= r) {
        // If the lightest and heaviest person can share a boat
        if (people[l] + people[r] <= limit) {
            // Increment the result
            result++;
            // Move the left pointer to the next person
            l++;
        }
        // If the heaviest person needs to take a boat alone
        else if (people[r] <= limit || l === r && people[l] <= limit) {
            // Increment the result
            result++;
        }
        // Move the right pointer to the next person
        r--;
    }
    // Return the minimum number of rescue boats required
    return result;
};

File: c/leet_code/881.c
/**
 * @file 881.c
 * @brief This module provides functions to calculate the minimum number of boats required to rescue people.
 * @author JerodG <https://github.com/jerodg>
 * @date 04May24
 */

#define OPTIMIZE_LEVEL "O3,unroll-loops"

#pragma GCC optimize(OPTIMIZE_LEVEL)
#include <stdlib.h>

/**
 * @brief Compares two integers.
 * @param a The first integer to compare.
 * @param b The second integer to compare.
 * @return The difference between the two integers.
 */
int compare(const void* a, const void* b) { return (*(int*)b - *(int*)a); }

/**
 * @brief Sorts an array of integers in descending order.
 * @param array The array to sort.
 * @param size The size of the array.
 */
void sort_integers(int* array, const unsigned int size) { qsort(array, size, sizeof(int), compare); }

/**
 * @brief Calculates the minimum number of boats required to rescue people.
 * @param people An array of integers where each integer represents the weight of a person.
 * @param peopleSize The size of the people array.
 * @param limit The maximum weight that a boat can carry.
 * @return The minimum number of boats required.
 *
 * This function first sorts the people array in descending order. Then, it iterates over the array from both ends,
 * trying to fit the heaviest and lightest person in the same boat. If they can't fit together, the heaviest person
 * takes a boat alone. The process continues until all people have been assigned to a boat.
 */
int numRescueBoats(int* people, const int peopleSize, const int limit) {
    sort_integers(people, peopleSize);

    int num_boats = 0;                 // The number of boats required
    int heavy_person = 0;              // Index of the heaviest person who has not yet been assigned to a boat
    int light_person = peopleSize - 1; // Index of the lightest person who has not yet been assigned to a boat

    while (heavy_person <= light_person) {
        if (people[heavy_person] + people[light_person] <= limit) {
            num_boats++;
            heavy_person++;
            light_person--;
        } else {
            num_boats++;
            heavy_person++;
        }
    }
    return num_boats;
}
